[dev] Tore out dropdown button out of dropdown

morethanfire opened the pull request:
This is 1 of 2 tasks to split apart the Dropdown component.

groberts314 commented on docs/src/atoms/dropdown.js:
```diff
@@ -1,15 +1,25 @@
+ * NOTE: The props 'button' part of this component's documentation has been is deprecated as of Feb. 11, 2020.
+ * Please use the DropdownButton component instead.
+ * We are disabling Eslint for this file because of the above note and wanting to
+ * deprecate it fully in hopes to rip out the select portion of it as well.
```
Can I better understand why it is necessary to completely disable ESLINT?

morethanfire replied:
Because I don't see a need to fix the whole entire file when it's about to be fully deprecated.

groberts314 commented on docs/src/molecules/dropdownButton.jsx:
```diff
@@ -0,0 +1,61 @@
+            name: 'iconType',
+            type: 'chevron-down<br />| plus',
+            default: 'chevron-down',
+            description: 'An element type to render as.',
```
This description seems off for `iconType` prop (maybe it was accidentally copied and pasted from an `as` prop somewhere...?)

morethanfire replied:
Oh, you're right. I forgot to update it.

morethanfire replied:
fixed.

groberts314 commented on src/molecules/dropdownButton.jsx:
```diff
@@ -0,0 +1,107 @@
+
+const propTypes = {
+    ...buttonPropTypes,
+    as: PropTypes.oneOf(['button']),
```
Why keep this is a `oneOf` (or even have it at all) if the only valid value is "button"?

morethanfire replied:
You're right.

morethanfire replied:
fixed.

morethanfire commented on docs/src/atoms/buttonIconExample.jsx:
```diff
@@ -0,0 +1,23 @@
+    return (
+        <div>
+            <Button icon>
+                <Icon color="inverse" type="plus" />
```
Makes sense.

morethanfire commented on docs/package.json:
```diff
@@ -5,6 +5,8 @@
     "build:clean": "rimraf build",
     "start": "webpack-dev-server --mode development --content-base docs/build/ --hot --progress --colors --host 0.0.0.0",
+    "start:local": "npm run start",
+    "start:local:prod": "npm run build",
```
I want to do more with the script that just build, but kind of what we do over in HC.

morethanfire replied:
Fixing.

morethanfire commented:
I added a few more commits to this PR before I wanted to call it done. Apologies for that.

Added sense last review:
- A component
- Prompt's children renderer was reworked in order to work with any component. Thinking about DropdownButton.
- Color import fixes.
- Docs Dropdown button prop components were replaced with DropdownButton.
- A few misc. fixes around docs.

groberts314 commented on docs/src/atoms/promptDropdownButtonExample.jsx:
```diff
@@ -0,0 +1,74 @@
+                    id="delete"
+                    label="Delete"
+                    onClick={() => onPromptClick('delete')}
+                />
```
For the sake of the docs showing good coding practice, is it better if this becomes a class component, and has separate, bound functions for handling "delete" versus "edit", so as to avoid the anonymous functions in the render?  Just a thought.  NAB.

morethanfire replied:
According to the React docs writing function components is the way they're want everyone to head. I think here, https://reactjs.org/docs/hooks-state.html

groberts314 commented on docs/src/atoms/promptExample.jsx:
```diff
@@ -0,0 +1,26 @@
+            <br />
+
+            <Prompt inline>
+                <A>asdf</A>
```
Should this say something other than "asdf"?  "Click Me"?

morethanfire replied:
Thanks!

groberts314 commented on docs/src/atoms/promptExample.jsx:
```diff
@@ -0,0 +1,26 @@
+} from 'react-cm-ui';
+import React from 'react';
+
+function PrompExample() {
```
Misspelled `PromptExample` here (missing the "t" in "Prompt").

morethanfire replied:
Thanks

groberts314 commented on src/atoms/a.jsx:
```diff
@@ -0,0 +1,51 @@
+            style={style}
+        >
+            {children}
+        </span>
```
Very interesting... 🤔 
Are there any consequences (good or bad) to these actually being `<span>`s, rather than bona fide, 100% organic, good ol' fashioned `<a>` elements?

morethanfire replied:
It mostly has to do with people that have special needs, but as well the docs say not using an `href` with an a tag isn't semantically correct.

groberts314 replied:
Got it.  Yeah, thinking about this more, basically the warnings are telling us that we're kind of misusing `<a>` elements.  So, our response is, okay then, we won't use `<a>` elements for that; we'll use `<span>` elements with `role="button"` to declare that they're interactive elements.  Makes sense.

groberts314 commented on src/atoms/prompt.jsx:
```diff
@@ -71,45 +107,61 @@ class Prompt extends Component {
+    //         onClick: this.onClick,
+    //         ref: (promptAction) => { this.promptAction = promptAction; },
+    //     });
+    // }
```
Can we please remove the commented code, once we commit to the current implementation?

groberts314 commented on src/atoms/prompt.jsx:
```diff
@@ -140,60 +192,51 @@ class Prompt extends Component {
+                {/* {this.renderAction()} */}
+                <div
+                    onClick={this.onClick}
+                    onKeyDown={() => {}}
```
Should we have a `const noop = () => {};` outside of `render()` and use that instead?  I thought anonymous functions inside the `render()` was a no-no (essentially as bad as `.bind(this)` inside `render()`).

groberts314 commented on src/atoms/prompt.jsx:
```diff
@@ -140,60 +192,51 @@ class Prompt extends Component {
+                    <div
+                        className="prompt-yes-btn"
+                        onClick={this.onYesClick}
+                        onKeyDown={() => {}}
```
Another case for my `noop` const

groberts314 commented on src/atoms/prompt.jsx:
```diff
@@ -140,60 +192,51 @@ class Prompt extends Component {
+                    <div
+                        className="prompt-no-btn"
+                        onClick={this.onNoClick}
+                        onKeyDown={() => {}}
```
Another case for my `noop` const

groberts314 commented on src/templates/actionBar.jsx:
```diff
@@ -136,185 +164,217 @@ class ActionBar extends React.PureComponent {
+                                                    title={dropdownButton.title}
+                                                >
+                                                    {_.map(dropdownButton.options, (option) => {
+                                                        dropdownButtonOptionKeyNum += 1;
```
Look at you not using unary `++` and then suppressing the warning! 😁 
I think I'm perhaps a bit more of a rebel...

groberts314 commented on src/templates/actionBar.jsx:
```diff
@@ -136,185 +164,217 @@ class ActionBar extends React.PureComponent {
-                                                    );
-                                                })}
-                                            </List>
+                                            {list && _.isArray(list) && (
```
Is it necessary to do `list && _.isArray(list)`?  Shouldn't `_.isArray(list)` be sufficient?

groberts314 commented on src/templates/actionBar.jsx:
```diff
@@ -136,185 +164,217 @@ class ActionBar extends React.PureComponent {
+                                                    {_.map(list, (item) => {
+                                                        const {
+                                                            actionsButton,
+                                                            jsx, // eslint-disable-line no-shadow
```
Hmm, are we sure this is okay?  This is one where (despite my rebelliousness mentioned above) I'd probably heed the warning and maybe alias this as `innerJsx` or some such to avoid the variable shadowing.  That's one of those things in my book where the warning is correct in telling you you might be doing something dangerous; think twice.  If it's correct, and you want to leave it as-is, no objections.  NAB.

morethanfire replied:
Oh, let me get this one.

groberts314 commented on src/templates/actionBar.jsx:
```diff
@@ -136,185 +164,217 @@ class ActionBar extends React.PureComponent {
+                                                            'action_bar--list_item-icon_table': iconTable,
+                                                        });
+
+                                                        const itemKey = `action_bar--list_item-${gridColumnListItemKeyNum++}`; // eslint-disable-line no-plusplus
```
Haha there it is! 😁 `++` FTW!

groberts314 replied:
`++` has been in so many programming languages using syntax for so long, I'm not sure when it officially because "considered harmful".  I think there are situations where you should probably avoid it (i.e. incrementing a variable and reading its value at the same time is just not smart there) but I think banning it wholesale is a bit harsh and unnecessary.

groberts314 commented on src/atoms/prompt.jsx:
```diff
@@ -140,60 +146,51 @@ class Prompt extends Component {
+                id={id}
+                style={style}
+            >
+                {/* {this.renderAction()} */}
```
Is there a reason to keep this guy around?  NAB of course.
